### PR TITLE
Validate XML mapping against XSD file

### PR DESCRIPTION
**Pull request by greg0ire**
# TODO

- [x] fix all the tests I broke
- [x] make it optional and disabled by default, and trigger deprecation if disabled?
- [x] establish what is the right kind of exception to throw (there is a test that asserts a mapping exception from doctrine/common should be thrown, I don't know why yet)

### CODE REVIEW

File: lib/Doctrine/ORM/Mapping/MappingException.php
Comment (greg0ire) [inline lib/Doctrine/ORM/Mapping/MappingException.php]: First time ever playing with `array_reduce`, this is fun! But I should probably use implode instead, right?

Comment (alcaeus) [reply lib/Doctrine/ORM/Mapping/MappingException.php]: Yeah, that looks like it could perfectly be implemented with `implode(PHP_EOL, array_map(...);`.

Comment (greg0ire) [reply lib/Doctrine/ORM/Mapping/MappingException.php]: Will change

**Comment by greg0ire** [comment]
Regarding the exception issue, I found that the xml driver extends a class from another package (doctrine/common). [One of the parent methods throws an exception *from that packge* if the result from `loadMappingFile` is missing a `$className` key](https://github.com/doctrine/common/blob/a3e240fa07ec9748387c13534949d543c0e177fa/lib/Doctrine/Common/Persistence/Mapping/Driver/FileDriver.php#L115-L118).

Would you consider throwing an exception from this package a BC-break or not? Also, if we forget about this PR for a moment, wouldn't it be better to do this:

```
<?php
try {
   // method call from another package
} catch (ExceptionFromAnotherPackage $e) {
    throw new ExceptionFromThisPackage('message', 0, $e);
}
```

File: lib/Doctrine/ORM/Mapping/Driver/XmlDriver.php
Comment (Majkl578) [inline lib/Doctrine/ORM/Mapping/Driver/XmlDriver.php]: How about call it `validateMapping`?

File: lib/Doctrine/ORM/Mapping/Driver/XmlDriver.php
Comment (Majkl578) [inline lib/Doctrine/ORM/Mapping/Driver/XmlDriver.php]: This will cause duplicated IO, better pass already fetched XML string to this function and use `DOMDocument::loadXML()`.

File: lib/Doctrine/ORM/Mapping/Driver/XmlDriver.php
Comment (Majkl578) [inline lib/Doctrine/ORM/Mapping/Driver/XmlDriver.php]: These two directives could be put into a more DRY `finally` block, and use `$backedUpErrorSetting`.

File: lib/Doctrine/ORM/Mapping/Driver/XmlDriver.php
Comment (Majkl578) [inline lib/Doctrine/ORM/Mapping/Driver/XmlDriver.php]: Space around `!` is needed.

**Review by Majkl578 — CHANGES_REQUESTED** [review]
I think this would be a nice addition to 3.0 and have it enabled by default.

_Note that this introduces a dependency on ext-dom._

**Comment by Majkl578** [comment]
> Would you consider throwing an exception from this package a BC-break or not?

In 3.0: No.
In 2.x: No, because this behavior would not be enabled by default (if accepted for 2.x - not sure about it).

---

Please add some tests to illustrate how the error propagation from libxml looks like.
It'd also be interesting to know how many (if any) tests fail with this enabled.

File: lib/Doctrine/ORM/Mapping/Driver/XmlDriver.php
Comment (greg0ire) [reply lib/Doctrine/ORM/Mapping/Driver/XmlDriver.php]: Could be less misleading since I don't use actual assertions here

File: lib/Doctrine/ORM/Mapping/Driver/XmlDriver.php
Comment (greg0ire) [reply lib/Doctrine/ORM/Mapping/Driver/XmlDriver.php]: Yeah but it gives worse error messages

**Comment by greg0ire** [comment]
Right now it's not optional so... congrats, no failures!

**Comment by greg0ire** [comment]
> Right now it's not optional so... congrats, no failures!

Uuuuh what am I saying, there is one failure that was not intended (I think):

```
1) Doctrine\Tests\ORM\Tools\Export\XmlClassMetadataExporterTest::testExportDirectoryAndFilesAreCreated
Doctrine\ORM\Mapping\MappingException: libxml error: Element '{http://doctrine-project.org/schemas/orm/doctrine-mapping}join-column', attribute 'on-update': The attribute 'on-update' is not allowed.
/home/greg/dev/doctrine2/tests/Doctrine/Tests/ORM/Tools/Export/xml/Doctrine.Tests.ORM.Tools.Export.User.dcm.xml:35
libxml error: Element '{http://doctrine-project.org/schemas/orm/doctrine-mapping}cascade-persist': This element is not expected. Expected is one of ( {http://doctrine-project.org/schemas/orm/doctrine-mapping}cascade-detach, ##other{http://doctrine-project.org/schemas/orm/doctrine-mapping}* ).
/home/greg/dev/doctrine2/tests/Doctrine/Tests/ORM/Tools/Export/xml/Doctrine.Tests.ORM.Tools.Export.User.dcm.xml:59

/home/greg/dev/doctrine2/lib/Doctrine/ORM/Mapping/MappingException.php:829
/home/greg/dev/doctrine2/lib/Doctrine/ORM/Mapping/Driver/XmlDriver.php:873
/home/greg/dev/doctrine2/lib/Doctrine/ORM/Mapping/Driver/XmlDriver.php:840
/home/greg/dev/doctrine2/vendor/doctrine/common/lib/Doctrine/Common/Persistence/Mapping/Driver/FileDriver.php:115
/home/greg/dev/doctrine2/lib/Doctrine/ORM/Mapping/Driver/XmlDriver.php:59
/home/greg/dev/doctrine2/lib/Doctrine/ORM/Mapping/ClassMetadataFactory.php:151
/home/greg/dev/doctrine2/vendor/doctrine/common/lib/Doctrine/Common/Persistence/Mapping/AbstractClassMetadataFactory.php:333
/home/greg/dev/doctrine2/lib/Doctrine/ORM/Mapping/ClassMetadataFactory.php:78
/home/greg/dev/doctrine2/vendor/doctrine/common/lib/Doctrine/Common/Persistence/Mapping/AbstractClassMetadataFactory.php:226
/home/greg/dev/doctrine2/vendor/doctrine/common/lib/Doctrine/Common/Persistence/Mapping/AbstractClassMetadataFactory.php:115
/home/greg/dev/doctrine2/tests/Doctrine/Tests/ORM/Tools/Export/AbstractClassMetadataExporterTest.php:91
```

File: lib/Doctrine/ORM/Mapping/Driver/XmlDriver.php
Comment (greg0ire) [reply lib/Doctrine/ORM/Mapping/Driver/XmlDriver.php]: See https://github.com/doctrine/doctrine2/pull/6728#issuecomment-333406605

**Comment by greg0ire** [comment]
> Note that this introduces a dependency on ext-dom.

Made that explicit

File: lib/Doctrine/ORM/Mapping/Driver/XmlDriver.php
Comment (greg0ire) [reply lib/Doctrine/ORM/Mapping/Driver/XmlDriver.php]: I lost my train of thought it seems

File: composer.json
Comment (SenseException) [inline composer.json]: I have a question: Isn't this extension optional? The usage of XML in Doctrine is optional too for a user, isn't it?

Comment (alcaeus) [reply composer.json]: Technically, it's an optional dependency. However, making it an explicit dependency is the defensive way here: we prevent the code from blowing up in the users face with a fatal error if they use XML mappings.

Remember that other bundles (e.g. the FOSUserBundle) specify XML mappings which are automatically imported, so the user won't know that he's using XML mappings and thus should add the `ext-dom` dependency.

One more thought: moving this to `suggest` would require all bundles that specify XML mappings to add a `require` entry for `ext-dom`, which would be a BC break towards the bundles. IMO it's safer to add it as an explicit dependency.

Comment (SenseException) [reply composer.json]: Talking about "Bundles" sounds more like Symfony instead of Doctrine. This repository is independent from frameworks. Should in this case the `require` be added in DoctrineBundle? What are "all bundles"?

Comment (alcaeus) [reply composer.json]: @SenseException replace with "bundles" with "anything that provides XML mappings". Suggesting `ext-dom` instead of requiring it will break third party software if they already use XML mappings. That's a BC break that would move this feature back to 3.0. The way it is now it's possible to add this to the 2.x line.

Comment (greg0ire) [reply composer.json]: @alcaeus I rebased on 3.x yesterday :sweat_smile:

Comment (alcaeus) [reply composer.json]: Oh, didn't see that, my bad. IMO it's better to have the explicit dependency anyways. 👍

**Comment by Majkl578** [comment]
Needs a rebase.

File: composer.json
Comment (Majkl578) [reply composer.json]: I have mixed feelings about this. One one hand, it's true that user would need explicit ext-dom dependency. On the other one, we should check at runtime for DOM presence, it just feels wrong to force people with annotations install DOM. (Well, we can say the same about doctrine/annotations, but that's slightly different story).
Since this PR is now 3.0 only, it should be fine fine to do either way, but question is which one is better.

Comment (Ocramius) [reply composer.json]: > On the other one, we should check at runtime for DOM presence, it just feels wrong to force people with annotations install DOM.

https://webmozart.io/blog/2014/10/24/defining-php-annotations-in-xml/

**Comment by greg0ire** [comment]
>  establish what is the right kind of exception to throw (there is a test that asserts a mapping exception from doctrine/common should be thrown, I don't know why yet)

#7199 should solve this

**Review by lcobucci — CHANGES_REQUESTED** [review]
@greg0ire @Majkl578 I've rebased this branch and adjusted the files that were actually invalid according to our XSD. The only broken test we have now are about the validation of an invalid XML document (which was being tested in a different way).

About the design of the solution, this implementation describes EXACTLY the point I was trying to make regarding deprecations. IMO we **should not** modify the existing class but rather create a new one which validates the XML files and simply add the `@deprecated` annotation on the existing one. This new class can be a simple decorator for now.

We should also consider that the current approach breaks [OCP](https://web.archive.org/web/20060822033314/http://www.objectmentor.com/resources/articles/ocp.pdf)...

**Comment by lcobucci** [comment]
BTW I do believe that this would fit nicely for v2.7 and I don't really see a reason for blocking this until we decide about #7199

**Comment by greg0ire** [comment]
> simply add the @deprecated annotation on the existing one. 
> This new class can be a simple decorator for now.

This would create a situation where you have an unfixable deprecation.

> BTW I do believe that this would fit nicely for v2.7 and I don't really see a reason for blocking this until we decide about #7199

I use `throw MappingException` from ORM, but it does not extend `MappingException` from Common, and should.

**Comment by lcobucci** [comment]
> This would create a situation where you have an unfixable deprecation.

Nope, the migration path is to use the new class. Users would only be notified about the deprecation when using directly the old class. Nobody should be bothered if we're using internally our own deprecated items (this already happens with `ClassMetadata`		and `ClassMetadataInfo` btw).

**Comment by greg0ire** [comment]
> Nope, the migration path is to use the new class. Users would only be notified about the deprecation when using directly the old class. Nobody should be bothered if we're using internally our own deprecated items (this already happens with ClassMetadata	and ClassMetadataInfo btw).

Somewhat agree, but are there any tools that are capable of detecting calls like you describe and only act on direct calls? phpstan-deprecation-rules, maybe?

**Comment by greg0ire** [comment]
> (this already happens with ClassMetadata	and ClassMetadataInfo btw

Is this why https://github.com/doctrine/doctrine2/pull/6886 never got merged?

**Comment by lcobucci** [comment]
> Somewhat agree, but are there any tools that are capable of detecting calls like you describe and only act on direct calls? phpstan-deprecation-rules, maybe?

That's correct, that PHPStan extension behaves like that indeed. But please keep in mind that my main concern is about our design principles and values... IMO we can always adapt tools but sticking to what we believe is the most appropriated is usually harder - specially on these days we live where we're always eager to push anything to production (definitely not talking about the code here, just a rant of an old grumpy developer HAHAHA).

Anyway, sorry if this was way philosophical :smile:

**Comment by Kern046** [comment]
Hello there, I volunteered in #9637 to help in this PR which seems to have mobilized some energy for some time now ! As a first contribution it also seems quite tough viewing the discussions, so I will rely on your advice to make it progress properly !

My first question for now is about the target branch, I saw that it was targeting 2.10.x a few months ago, should it target 2.11.x now ? I also noticed [here](https://github.com/doctrine/orm/pull/6728#discussion_r142091148) that the branch was rebased with 3.x, but if I rebase it now shouldn't I use 2.11.x instead ?

Thanks a lot :)

**Comment by greg0ire** [comment]
Hi! Thanks for volunteering! I wasn't sure if you wanted to do it so I did some cleanup. I retargeted to 2.12.x, which is the next minor branch. You can start from there :)

**Comment by greg0ire** [comment]
Apparently I broke the tests in `tests/Doctrine/Tests/ORM/Mapping/XmlMappingDriverTest.php`, it seems that the mapping files (one of which is now invalid) were used for more than one test. And maybe what I put in the message of `markTestSkipped` is wrong.

**Comment by Kern046** [comment]
I did a PR to your branch with the following elements :

* Restoration of the removed DDC889 XML mapping files which made a test fail
* Implementation of new tests for incorrect cases (invalid and missing attributes)

Though I have several concerns about this PR :

* Currently we have to retrieve the XSD file through `www.doctrine-project.org`, so I wonder : what if someday this website is unavailable for some reason or encounters heavy latencies ? Won't this have a strong impact on everyone implementing XML mapping ?
* In the PR tests we seem unable to reach `XmlDriver::validateMapping` to ensure the whole workflow works as intended. The tests use directly `DOMDocument` and doesn't produce the same error output as the driver which uses `libxml_errors`. Due to that, I think that the current tests do not succeed in their objective of protecting `XmlDriver::validateMapping` against regressions or misbehavior.

WDYT ?

**Comment by greg0ire** [comment]
> Currently we have to retrieve the XSD file through www.doctrine-project.org, so I wonder : what if someday this website is unavailable for some reason or encounters heavy latencies ? Won't this have a strong impact on everyone implementing XML mapping ?

Where/when does that happen exactly? Are you sure it happens? Aren't we avoiding that with this line: https://github.com/doctrine/orm/pull/6728/files#diff-3613bafacbade59e06613442d37c5f0debfed82f00f548143197ed7caf98a635R978

> In the PR tests we seem unable to reach XmlDriver::validateMapping to ensure the whole workflow works as intended. The tests use directly DOMDocument and doesn't produce the same error output as the driver which uses libxml_errors. Due to that, I think that the current tests do not succeed in their objective of protecting XmlDriver::validateMapping against regressions or misbehavior.

:thinking: don't we reach it through `testInvalidEntityOrMappedSuperClassShouldMentionParentClasses`? If yes, let's do something similar to that?

**Comment by Kern046** [comment]
> Where/when does that happen exactly? Are you sure it happens? Aren't we avoiding that with this line: https://github.com/doctrine/orm/pull/6728/files#diff-3613bafacbade59e06613442d37c5f0debfed82f00f548143197ed7caf98a635R978

Oh yeah absolutely, I assumed that we had to read the XML mapping file XSD location and then access it remotely. Problem solved :smile: !

> thinking don't we reach it through `testInvalidEntityOrMappedSuperClassShouldMentionParentClasses`? If yes, let's do something similar to that?

Very nice ! I adapted the tests, will push it tomorrow once it's all green !

**Comment by Kern046** [comment]
I submitted the tests improvement to your repo branch. Let me know what you think about it :)

**Comment by greg0ire** [comment]
You can open your own PR to `doctrine/orm` if you want to iterate faster. Anyway, this is starting to look good, and we should think how we could make this optional and disabled by default. I guess we should probably add a new argument to `ORMSetup::createXMLMetadataConfiguration`, set to `false` by default,  deprecate not setting it to `true`, and remove it in 3.0?

**Comment by Kern046** [comment]
> You can open your own PR to `doctrine/orm` if you want to iterate faster. Anyway, this is starting to look good, and we should think how we could make this optional and disabled by default. I guess we should probably add a new argument to `ORMSetup::createXMLMetadataConfiguration`, set to `false` by default, deprecate not setting it to `true`, and remove it in 3.0?

If you don't mind I think it's better to keep the current PR to keep its history ? I may be wrong on this, this is my first contribution to Doctrine so if you think it's better to open my own PR I will :).

Okay for the optional design, I'll give it a try asap.

**Comment by greg0ire** [comment]
> If you don't mind I think it's better to keep the current PR to keep its history ? I may be wrong on this, this is my first contribution to Doctrine so if you think it's better to open my own PR I will :).

If you don't mind making PRs to my repo, I'm fine with it, this was a suggestion for your convenience. You somehow manage to get green tests everytime so it should be fine :)

File: composer.json
Comment (greg0ire) [inline composer.json]: @Kern046 this needs to be moved to `suggest`, and then if it's not loaded, validation should be skipped.

Comment (Kern046) [reply composer.json]: Should it also trigger a deprecation or not ? If not, that means that people could have XSD validation enabled with the boolean, but ineffective due to the absence of the DOM extension.

Comment (greg0ire) [reply composer.json]: Gave if some thought, and I think it should, and then we throw in 3.0, otherwise people are going to report issues that could have been prevented by installing that extension. In fact, I think you can even throw an exception rightaway: it's not like people will get the exception by default.

**Review by SenseException — DISMISSED** [review]

File: lib/Doctrine/ORM/Mapping/Driver/XmlDriver.php
Comment (derrabus) [inline lib/Doctrine/ORM/Mapping/Driver/XmlDriver.php]: These lines should be moved into the `try` block, just in case.

File: lib/Doctrine/ORM/Mapping/MappingException.php
Comment (derrabus) [inline lib/Doctrine/ORM/Mapping/MappingException.php]: Don't we need some kind of separator between message and file?

File: tests/Doctrine/Tests/ORM/Mapping/XmlMappingDriverTest.php
Comment (derrabus) [inline tests/Doctrine/Tests/ORM/Mapping/XmlMappingDriverTest.php]:
```suggestion
        } catch (MappingException $throwable) {
```

Additionally, we need to make sure the test fails if no exception is raised at all.

**Review by derrabus — CHANGES_REQUESTED** [review]

**Review by beberlei — DISMISSED** [review]
An unrelated follow up could be to get rid of SimpleXML in the driver and use only ext-dom

**Review by derrabus — DISMISSED** [review]

**Comment by greg0ire** [comment]
@lcobucci regarding the OCP, I read the link you gave and it says (rightly IMO) that the OCP is something that should be taken into account when designing a class. `XMLDriver` does not seem closed against adding validation functionality, as there is no way to inject a validator in it. Also, I agree that decoration is a great way to add functionality, and use it a lot in my projects, but here the method we would want to decorate is `loadMappingFile()`, which is protected. That means we cannot use decoration, but would be forced to use inheritance instead to "decorate" the method. Even if it was public and we could do proper decoration, the interface of `XMLDriver` is not exactly small, which means we would have to proxify a lot of methods to achieve that.

I think a sensible way to violate the OCP as little as possible could be to do the following:

```php
interface XMLValidator
{
    public function __invoke(string $file);
}

class DomXMLValidator implements XMLValidator
{
    // … implementation goes here
}

class XMLDriver
{
    // inject validator in constructor and call it
}
```

But given the size of the added piece of code, and given how unlikely I think it is people would want to inject their own implementation of the XMLValidator, I'm not sure it's worth it.

Do you still have concerns about runtime deprecations, now that we have switched to https://github.com/doctrine/deprecations/ ? That lib wasn't discussed at all in https://github.com/doctrine/doctrine-website/issues/200

**Comment by lcobucci** [comment]
> @lcobucci regarding the OCP, I read the link you gave and it says (rightly IMO) that the OCP is something that should be taken into account when designing a class. `XMLDriver` does not seem closed against adding validation functionality, as there is no way to inject a validator in it. (...)

The great thing about "principles" that version of me from 4 years ago hadn't learned well is that they aren't absolute rules :laughing:. I (now) know that understanding the context and allowing some flexibility is important - even though I tend to stick to them as much as possible.

If I was making this feature, I'd still aim to add a new component instead of modifying the existing one to conditionally change its behaviour based on a boolean flag. I believe that adding a new (`final`) class (`SchemaEnforcingXmlDriver`, whateva) that - initially - inherits the `XmlDriver` makes things more explicit in the long run and gives more flexibility on refactoring than the boolean flag approach :smile:

With that said, I understand your arguments and agree that a completely different design would make things much better buuuuut :rofl:


> Do you still have concerns about runtime deprecations, now that we have switched to https://github.com/doctrine/deprecations/ ? That lib wasn't discussed at all in [doctrine/doctrine-website#200](https://github.com/doctrine/doctrine-website/issues/200)

Doctrine has fully embraced that deprecations library so I believe that's a moot point now :+1:

File: tests/Doctrine/Tests/ORM/Mapping/XmlMappingDriverTest.php
Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Mapping/XmlMappingDriverTest.php]: This doesn't seem correct =)

https://phpstan.org/r/7db17906-8a40-4f4e-a45d-16ec59ef3f8d on the `non-empty-array` :+1: 

```suggestion
     * @param class-string $class
     * @param non-empty-array<string, int> $expectedExceptionOccurrences
```

File: tests/Doctrine/Tests/ORM/Mapping/XmlMappingDriverTest.php
Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Mapping/XmlMappingDriverTest.php]: This doesn't look correct (https://phpstan.org/r/3986d513-0d9f-4591-9d33-6a9fae6ec4b8 and https://psalm.dev/r/1d25f7f2f6).

I'd also just use `@return`, IDEs are much better nowadays...

```suggestion
     * @return non-empty-list<array{0: class-string, 1: non-empty-array<string, int>}>
```

**Review by lcobucci — DISMISSED** [review]
Just some minor improvements on the SA side of things =)

**Comment by greg0ire** [comment]
Thanks for reviving this @Kern046 ! And thanks for your help @lcobucci ! 5 years after the start, let's finally merge this :)

**Pull request merged**
